### PR TITLE
first revision of persisent index storage for downsample filodb clusters

### PR DESCRIPTION
**Pull request by kvpetrov**
**Pull Request checklist**

- [ ] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [ ] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

**Current behavior :** (link exiting issues here : https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)

**New behavior :**

**BREAKING CHANGES**

If this PR contains a breaking change, please describe the impact and migration
path for existing applications.
If not please remove this section.

Breaking changes may include:
- Any schema changes to any Cassandra tables
- The serialized format for Dataset and Column (see .toString methods)
- Over the wire formats for Akka messages / case classes
- Changes to the HTTP public API
- Changes to query parsing / PromQL parsing

**Other information**:

### CODE REVIEW

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala [hunk 1]
```diff
@@ -204,12 +209,17 @@ class DownsampledTimeSeriesShard(rawDatasetRef: DatasetRef,
         stats.indexEntriesRefreshed.increment(count)
         logger.info(s"Refreshed downsample index with new records numRecords=$count " +
           s"dataset=$rawDatasetRef shard=$shardNum fromHour=$fromHour toHour=$toHour")
+        partKeyIndex.commit()
```
Comment (amolnayak311) [inline core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala:212]: I am not sure ``commit`` guarantees sync to file system. We will have to look into Lucene internals.

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala [hunk 2]
```diff
@@ -204,12 +209,17 @@ class DownsampledTimeSeriesShard(rawDatasetRef: DatasetRef,
...
+        val hourMillis = toHour  * 1000 * 60 * 60
```
Comment (amolnayak311) [inline core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala:213]: It would be better to get the max time we pulled from the underlying store to avoid missed data.

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala [hunk 3]
```diff
@@ -204,12 +209,17 @@ class DownsampledTimeSeriesShard(rawDatasetRef: DatasetRef,
...
       }
       .onErrorHandle { e =>
         stats.indexRefreshFailed.increment()
         logger.error(s"Error occurred when refreshing downsample index " +
           s"dataset=$rawDatasetRef shard=$shardNum fromHour=$fromHour toHour=$toHour", e)
       }
+
   }
 
   private def startStatsUpdateTask(): Unit = {
```
Comment (amolnayak311) [inline core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala:225]: We will also need to sync from ``indexRefresh`` periodically

**Review by amolnayak311 — CHANGES_REQUESTED** [review]

File: core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala (hunk 1 above)
Comment (amolnayak311) [reply core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala:212]: EDIT: I think commit looks ok at least from the documentation 


https://lucene.apache.org/core/7_4_0/core/org/apache/lucene/index/IndexWriter.html#commit--

File: core/src/test/scala/filodb.core/memstore/DownsampleIndexCheckpointerSpec.scala [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    val baseName = s"DownsampleIndexCheckpointerSpec-${System.currentTimeMillis()}-"
+    val tempDir = new File(baseDir, baseName)
+    tempDir.mkdir()
+    //val c = new DownsampleIndexCheckpointer();
+    val time = DownsampleIndexCheckpointer.getDownsampleLastCheckpointTime(tempDir)
+    time shouldEqual 0
+    DownsampleIndexCheckpointer.writeCheckpoint(tempDir, 10)
+    val time2 = DownsampleIndexCheckpointer.getDownsampleLastCheckpointTime(tempDir)
+    time2 shouldEqual 10
+  }
```
Comment (amolnayak311) [inline core/src/test/scala/filodb.core/memstore/DownsampleIndexCheckpointerSpec.scala:21]: Can we add cases when file is garbled/not in expected format? Also you test both cases directory empty and reading a value you wrote. Can this be split in to tests for readability ?

Thanks again for all the changes.

File: core/src/main/scala/filodb.core/downsample/DownsampleConfig.scala [hunk 5]
```diff
@@ -29,6 +32,15 @@ final case class DownsampleConfig(config: Config) {
   val schemas = if (config.hasPath ("raw-schema-names")) config.as[Seq[String]]("raw-schema-names")
                 else Seq.empty
 
+  val indexLocation = config.getOrElse[Option[String]]("index-location", None)
+
+  def indexLocationFile(ref: DatasetRef, shardNum: Int): File = {
+    indexLocation match {
+      case Some(il) => new File(il)
```
Comment (vishramachandran) [inline core/src/main/scala/filodb.core/downsample/DownsampleConfig.scala:39]: Shouldn't the location be {{indexLocation}} / {{datasetRefName}} / {{shardNum}}

File: core/src/test/scala/filodb.core/memstore/DownsampleIndexCheckpointerSpec.scala (hunk 4 above)
Comment (kvpetrov) [reply core/src/test/scala/filodb.core/memstore/DownsampleIndexCheckpointerSpec.scala:35]: fixed

Comment (amolnayak311) [reply core/src/test/scala/filodb.core/memstore/DownsampleIndexCheckpointerSpec.scala:35]: Thanks a lot

**Review by amolnayak311 — APPROVED** [review]
LGTM
